src/huggingface_hub/_training_logger.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains a Mixin to push training logs to the Hub."""
+import os
+from concurrent.futures import Future
+from typing import List, Optional, Union
+
+from .hf_api import create_repo, upload_folder
+
+
+class HFLoggerMixin:
+    """
+    Mixin class to add `push_to_hub` method to a logger class.
+
+    The idea is to log data locally and then push it to the Hub only when requested (e.g. once every few epochs).
+    This Mixin class is not specific to any logger class as long as it implements a `log_dir` property to return the
+    current experiment directory. In practice, it is compatible with lightning loggers (e.g. TensorBoardLogger) when
+    data is saved locally.
+
+    The `push_to_hub` method is asynchronous, meaning the upload action will be executed in a separate thread.
+    Warning: you must be careful not to modify the log files while they are being pushed. Make sure to push to the Hub
+    only when your script is not logging data anymore (e.g. at the end of an epoch).
+
+    Example:
+        ```py
+        from tensorboardX import SummaryWriter
+
+        class HFTensorBoardLogger(HFLoggerMixin, SummaryWriter):
+            @property
+            def log_dir(self) -> str:
+                return self.logdir
+
+        logger = HFTensorBoardLogger("training/results", repo_id="test_hf_logger", path_in_repo="tensorboard")
```

Comment:
Would it be weird not to assign a `log_dir`? I think it's confusing that we need to point both somewhere on the hub and somewhere on the disk. Shouldn't it write on a temporary disk instead? so that basically the hub acts as the source of truth?

Reply:
I have not thought of this use case before TBH but it makes total sense. At first I wanted the logger to be as close as possible to the default logger (i.e. write locally as well). But yes we can make `log_dir` optional and if not provided we create a temporary directory.

Reply:
@thomasw21 actually by default `SummaryWriter` creates a local directory under `/runs`. I will keep this behavior as it is instead of creating a temporary directory by default. If we really want a tmp dir in the future, we can always add it as a new feature.